packages/arb-token-bridge-ui/src/components/common/AddCustomChain.tsx

```diff
@@ -0,0 +1,267 @@
+      .map(chain => {
+        return {
+          ...chain,
+          // make sure chainID is numeric
+          chainID: Number(chain.chainID)
+        }
+      })
+  }
+
+  return []
+}
+
+function mapOrbitConfigToOrbitChain(data: OrbitConfig): ChainWithRpcUrl {
+  return {
+    chainID: data.chainInfo.chainId,
+    confirmPeriodBlocks: 150,
+    ethBridge: {
+      bridge: data.coreContracts.bridge,
+      inbox: data.coreContracts.inbox,
+      outbox: data.coreContracts.outbox,
+      rollup: data.coreContracts.rollup,
+      sequencerInbox: data.coreContracts.sequencerInbox
+    },
+    rpcUrl: data.chainInfo.rpcUrl,
+    explorerUrl: data.chainInfo.explorerUrl,
+    isCustom: true,
+    name: data.chainInfo.chainName,
+    partnerChainID: data.chainInfo.parentChainId,
+    retryableLifetimeSeconds: 604800,
+    nitroGenesisBlock: 0,
+    nitroGenesisL1Block: 0,
+    depositTimeout: 900000,
+    isArbitrum: true,
+    tokenBridge: {
+      l1CustomGateway: data.tokenBridgeContracts.l2Contracts.customGateway,
+      l1ERC20Gateway: data.tokenBridgeContracts.l2Contracts.standardGateway,
+      l1GatewayRouter: data.tokenBridgeContracts.l2Contracts.router,
+      l1MultiCall: data.tokenBridgeContracts.l2Contracts.multicall,
+      l1ProxyAdmin: data.tokenBridgeContracts.l2Contracts.proxyAdmin,
+      l1Weth: data.tokenBridgeContracts.l2Contracts.weth,
+      l1WethGateway: data.tokenBridgeContracts.l2Contracts.wethGateway,
+      l2CustomGateway: data.tokenBridgeContracts.l3Contracts.customGateway,
+      l2ERC20Gateway: data.tokenBridgeContracts.l3Contracts.standardGateway,
+      l2GatewayRouter: data.tokenBridgeContracts.l3Contracts.router,
+      l2Multicall: data.tokenBridgeContracts.l3Contracts.multicall,
+      l2ProxyAdmin: data.tokenBridgeContracts.l3Contracts.proxyAdmin,
+      l2Weth: data.tokenBridgeContracts.l3Contracts.weth,
+      l2WethGateway: data.tokenBridgeContracts.l3Contracts.wethGateway
+    }
+  }
+}
+
+export const AddCustomChain = () => {
+  const [customChains, setCustomChains] = useState<Chain[]>(
+    getCustomChainsFromLocalStorage()
+  )
+  const [chainJson, setChainJson] = useState<string>('')
+  const [error, setError] = useState<string | null>(null)
+  const [isDeleteConfirmation, setIsDeleteConfirmation] = useState(false)
+
+  function onAddChain() {
+    setError(null)
+
+    try {
+      const data = (
+        chainJson.trim()
+          ? JSON.parse(chainJson.replace(/[\r\n]+/g, ''))
+          : undefined
+      ) as OrbitConfig
+
+      if (!data) {
+        throw new Error('JSON input is empty.')
+      }
+
+      const customChain = mapOrbitConfigToOrbitChain(data)
+
+      if (!allowedParentChainIds.includes(Number(customChain.partnerChainID))) {
+        throw new Error(
+          `Invalid partnerChainID ${customChain.partnerChainID}. Only Arbitrum testnet parent chains are allowed.`
+        )
+      }
+
+      customChain.isCustom = true
+
+      saveCustomChainToLocalStorage(customChain)
```

Comment:
we should check the config against existing custom chains in the local storage to prevent duplication esp when some logic (i.e. wagmi chain) relies on only returning 1 custom chain using the chain id